### PR TITLE
Added a round-to-multiple decomposition for aten.mm/aten.bmm

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87965
* #87731

cc @jansel @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

speeds up timm_vision_transformer from 1.14x to 1.45x

### CODE REVIEW

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+    if b_offset != 0:
+        b = round_dim(b, rounding_factor, 1)
+        return torch.ops.aten.mm(a, b)[:, :-(rounding_factor - b_offset)]
+    elif k_offset != 0:
+        a = round_dim(a, rounding_factor, 1)
+        b = round_dim(b, rounding_factor, 0)
+        return torch.ops.aten.mm(a, b)
+    # elif a_offset != 0:
+    #     a = round_dim(a, rounding_factor, 0)
+    #     return torch.ops.aten.mm(a, b)[:-(rounding_factor - a_offset), :]
```
Comment (bertmaher) [inline torch/_inductor/decomposition.py:229]: Is there a reason not to apply this to the `a` dimension?

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+def should_pad_bench(a, b, op):
+    with no_dispatch():
+        a = torch.randn_like(a)
+        b = torch.randn_like(b)
+        no_pad = bench(lambda: op(a, b), display=False)
+        def round_up(x):
+            return (x + rounding_factor - 1) // rounding_factor * rounding_factor
+        a_pad = a.new_empty([round_up(i) for i in a.shape])
+        b_pad = b.new_empty([round_up(i) for i in b.shape])
+        pad = bench(lambda: op(a_pad, b_pad), display=False)
```
Comment (bertmaher) [inline torch/_inductor/decomposition.py:179]: This benchmark sort of assumes the concat is free, right?  Which, maybe it sometimes is, but is that universally true?  I guess if it's a win anyways, it's fine, I'm just curious.

File: torch/_inductor/decomposition.py [hunk 3]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+        # print(no_pad, pad)
+        return no_pad > pad + 100
+
+def should_pad_mm(a, b):
+    if all(x % rounding_factor == 0 for x in a.shape + b.shape):
+        return False
+    flops = a.shape[0] * a.shape[1] * b.shape[1]
+    bw = a.shape[0] * a.shape[1] + b.shape[0] * b.shape[1] + a.shape[0] * b.shape[1]
+    if flops <= bw * 64 or flops <= 1e9:
+        return False
```
Comment (bertmaher) [inline torch/_inductor/decomposition.py:190]: Do we need this heuristic if we also do autotuning?

File: torch/_inductor/decomposition.py [hunk 4]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
     else:
         return NotImplemented  # go directly to lowering
 
+import time
+def bench(f, name=None, iters=100, warmup=5, display=True):
+    for _ in range(warmup):
+        f()
+    torch.cuda.synchronize()
```
Comment (bertmaher) [inline torch/_inductor/decomposition.py:152]: We need to only do this stuff if CUDA is available.

File: torch/_inductor/decomposition.py [hunk 5]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+    begin = time.time()
```
Comment (bertmaher) [inline torch/_inductor/decomposition.py:153]: Maybe consider `torch.cuda.Event()` for timing?  (I'm not honestly sure how much it matters...)

File: torch/_inductor/decomposition.py [hunk 6]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+    #     return NotImplemented
+    b_offset = b.shape[2] % rounding_factor
+    a_offset = a.shape[1] % rounding_factor
+    k_offset = a.shape[2] % rounding_factor
+    if k_offset != 0:
+        a = round_dim(a, rounding_factor, 2)
+        b = round_dim(b, rounding_factor, 1)
+        return torch.ops.aten.bmm(a, b)
+    elif b_offset != 0:
+        should_pad_bench(a, b, torch.bmm)
```
Comment (bertmaher) [inline torch/_inductor/decomposition.py:247]: It looks like we're not using the result of `should_pad_bench` here?  (And actually it looks like it's not used anywhere... hmm.  Is this a stale version?

**Review by bertmaher — CHANGES_REQUESTED** [review]
This is awesome.   I'm continually amazed at how concisely transforms can be written using define-by-run.  And the perf results sound really promising too.

Seems like the diff needs a bit of cleanup before landing (removing commented code, etc.) and I have some comments/questions inline, so just requesting changes for that.  Let's get this shipped :-D

File: torch/_inductor/decomposition.py [hunk 7]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+        # print(a.shape, b.shape)
+        # print(no_pad, pad)
+        return no_pad > pad + 100
+
+def should_pad_mm(a, b):
+    if all(x % rounding_factor == 0 for x in a.shape + b.shape):
+        return False
+    flops = a.shape[0] * a.shape[1] * b.shape[1]
+    bw = a.shape[0] * a.shape[1] + b.shape[0] * b.shape[1] + a.shape[0] * b.shape[1]
+    if flops <= bw * 64 or flops <= 1e9:
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:189]: On V100 we shouldn't pad fp32

File: torch/_inductor/decomposition.py [hunk 8]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+        return False
+    return True
+
+
+def should_pad_bmm(a, b):
+    if all(x % rounding_factor == 0 for x in a.shape + b.shape):
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:195]: we shouldn't look at batch dim at all most likely?

File: torch/_inductor/decomposition.py [hunk 9]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+        return False
+    flops = a.shape[0] * a.shape[1] * a.shape[2] * b.shape[2]
+    bw = a.shape[0] * a.shape[1] * a.shape[2] + b.shape[0] * b.shape[1] * b.shape[2] + a.shape[0] * a.shape[1] * b.shape[2]
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:198]: bw is not used

File: torch/_inductor/decomposition.py [hunk 10]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+    if not should_pad_mm(a, b):
+        return NotImplemented
+    # if not should_pad_bench(a, b, torch.mm):
+    #     return NotImplemented
+    b_offset = b.shape[1] % rounding_factor
+    a_offset = a.shape[0] % rounding_factor
+    k_offset = a.shape[1] % rounding_factor
+    if b_offset != 0:
+        b = round_dim(b, rounding_factor, 1)
+        return torch.ops.aten.mm(a, b)[:, :-(rounding_factor - b_offset)]
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:222]: lol recursively applying decompositions is cool

File: torch/_inductor/decomposition.py (hunk 1 above)
Comment (ngimel) [reply torch/_inductor/decomposition.py:229]: If `a` is not transposed, it's outermost dimension doesn't matter (but we should check for a or b being transposed)

File: torch/_inductor/decomposition.py [hunk 11]
```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
     else:
         return NotImplemented  # go directly to lowering
 
+import time
+def bench(f, name=None, iters=100, warmup=5, display=True):
```
Comment (eellison) [inline torch/_inductor/decomposition.py:149]: we should be caching the results of bench.

File: torch/_inductor/decomposition.py (hunk 2 above)
Comment (Chillee) [reply torch/_inductor/decomposition.py:179]: Yeah, it’s not always free, and the slice is not always free either (makes the output non-contiguous). But generally, if there’s a significant improvement in matmul perf, it’s worth it.

File: torch/_inductor/decomposition.py (hunk 3 above)
Comment (Chillee) [reply torch/_inductor/decomposition.py:190]: No, I started off writing this heuristic, but had some issues with actually getting it to work, so just wrote the autotuner to see what a perfect heuristic looks like.

File: torch/_inductor/decomposition.py (hunk 6 above)
Comment (Chillee) [reply torch/_inductor/decomposition.py:247]: I just added this here so I could benchmark the impact of my heuristic haha.

File: torch/_inductor/decomposition.py (hunk 7 above)
Comment (Chillee) [reply torch/_inductor/decomposition.py:189]: We also shouldn’t pad if tf32 is off 🤔

**Comment by Chillee** [comment]
Superseded by https://github.com/pytorch/pytorch/pull/88709

**Pull request closed without merging**